### PR TITLE
Plugins: Updated Design with Search and Filtering

### PR DESCRIPTION
**Pull request by roundhill**
Continuation of PR #9471 that adds enhanced searching and filtering, along with a few bug fixes.

Created an all new Jetpack-only view for .com plugins, following this design:
artboard copy 17
and these new category names: Engagement, Security, Appearance, Writing.

The legacy files are still available, in the off chance this page will eventually show non-Jetpack plugins too.

Before|After
---|---
before|after2

### CODE REVIEW

File: client/lib/mixins/url-search/component.js [hunk 1]
```diff
@@ -0,0 +1,40 @@
+/**
+ * External dependencies
+ */
+import React from 'react';
+import assign from 'lodash/assign';
```
Comment (dmsnell) [inline client/lib/mixins/url-search/component.js]: We're using a Babel transform to improve loading from `lodash` and so should use `import { assign } from 'lodash'` instead.

Why are we importing `assign` anyway? We're using `Object.assign()` in Calypso and that is safe to use

File: client/lib/mixins/url-search/component.js [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+
+/**
+ * Internal dependencies
+ */
+import { doSearch, getSearchOpen } from './index';
+
+export default function( Component ) {
+	const componentName = Component.displayName || Component.name || '';
+
+	return React.createClass( {
```
Comment (dmsnell) [inline client/lib/mixins/url-search/component.js]: could still use `class` syntax here

```js
return class extends Component {

};
```

File: client/lib/mixins/url-search/component.js [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+		displayName: 'Searchable' + componentName,
+
+		getInitialState() {
+			return {
+				searchOpen: false
+			};
+		},
```
Comment (dmsnell) [inline client/lib/mixins/url-search/component.js]: …and if we use a class, we can set this property here like so…

```js
return class extends Component {
	state = {
		searchOpen: false,
	};
}
```

File: client/lib/mixins/url-search/component.js [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+
+		componentWillReceiveProps( nextProps ) {
+			if ( ! nextProps.search ) {
+				this.setState( {
+					searchOpen: false
+				} );
+			}
+		},
```
Comment (dmsnell) [inline client/lib/mixins/url-search/component.js]: do we only want to change the state if the search prop doesn't exist? what about turning it back on?

either way, here is a way we could shorten this up

```js
componentWillReceiveProps: ( { search } ) => ! search && this.setState( { searchOpen: false } );
```

File: client/lib/mixins/url-search/component.js [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+
+		render() {
+			const props = assign( {}, this.props, {
+				doSearch: doSearch.bind( this ),
+				getSearchOpen: getSearchOpen.bind( this )
+			} );
+			return React.createElement( Component, props );
+		}
```
Comment (dmsnell) [inline client/lib/mixins/url-search/component.js]: Facebook doesn't appear to want to see `createElement` either if using JSX.

```js
render() {
	return (
		<Component { 
			...this.props, 
			doSearch: this.doSearch,
			getSearchOpen: this.getSearch,
		} />
	);
}
```

Note that I've changed the `bind()`ing here and will instead move it up into the class. The reason is that when we do this on every render, we actually create new bindings and totally new functions/closures on every render, which we don't want.

**Comment by drw158** [comment]
Looks great. The only 2 things I can see, both of which can probably happen at a later time.

- The padding in the Jetpack plugin card is different than self-hosted plugin list. I suggest `padding: 24px 0 24px 24px;` for the top card, `padding: 16px 0 16px 16px;` for the other feature cards (this will add more room for the text).
- Change border color for hover style for the clickable cards. It's done with a shadow.

Would be cool to use css variables for this :)
```
.card {
    --box-shadow-color: palegoldenrod;
    box-shadow: 1px 2px 3px var(--box-shadow-color);
}

.card:hover {
    --box-shadow-color: goldenrod;
}
```

File: client/lib/mixins/url-search/component.js [hunk 6]
```diff
@@ -2,7 +2,6 @@
  * External dependencies
  */
 import React from 'react';
```
Comment (dmsnell) [inline client/lib/mixins/url-search/component.js]:
```js
import React, { Component } from 'react';

…

return class extends Component {
```

**Comment by dmsnell** [comment]
left a comment about the import syntax which is mostly style but would fit more with existing Calypso guides. other than that I am happy with this 👍

**Comment by vindl** [comment]
Could we merge https://github.com/Automattic/wp-calypso/pull/9524 into this branch, before merging it with master?

**Comment by drw158** [comment]
:shipit:

**Comment by dmsnell** [comment]
:shipit:

**Comment by roundhill** [comment]
:shipit:

**Comment by Copons** [comment]
:shipit:

**Pull request merged**